File: java/src/test/java/org/ikigaidigital/TimeDepositCalculatorTest.java
package org.ikigaidigital;

import java.util.Collections;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.within;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

public class TimeDepositCalculatorTest {
    @ParameterizedTest
    @CsvSource({
        "basic, 1234.56, 30, 1234.56",    // No interest <= 30 days
        "basic, 1234.56, 31, 1235.59",    // 1% interest: 1234.56 * 0.01 / 12 = 1.03
        "student, 1234.56, 30, 1234.56",  // No interest <= 30 days
        "student, 1234.56, 31, 1237.65",  // 3% interest: 1234.56 * 0.03 / 12 = 3.09
        "student, 1234.56, 366, 1234.56", // No interest >= 366 days
        "premium, 1234.56, 45, 1234.56",  // No interest <= 45 days
        "premium, 1234.56, 46, 1239.70",  // 5% interest: 1234.56 * 0.05 / 12 = 5.14
        "unknown, 1234.56, 33, 1234.56"   // No interest due to an unknown plan type
    })
    public void updateBalance_Test(String planType, double balance, int days, double expectedBalance) {
        TimeDepositCalculator calc = new TimeDepositCalculator();
        TimeDeposit deposit = new TimeDeposit(1, planType, balance, days);

        calc.updateBalance(Collections.singletonList(deposit));

        assertThat(deposit.getBalance()).isCloseTo(expectedBalance, within(0.001));
    }
}

File: kotlin/src/test/kotlin/org/ikigaidigital/TimeDepositCalculatorTest.kt
package org.ikigaidigital

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.math.RoundingMode

class TimeDepositCalculatorTest {

    @Test
    fun `updateBalance should calculate interest for basic plan`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(1, "basic", 1000.00, 45)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(BigDecimal(1000.83).setScale(2, RoundingMode.HALF_UP).toDouble())
    }

    @Test
    fun `updateBalance should calculate interest for student plan`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(2, "student", 1000.00, 365)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(BigDecimal(1002.50).setScale(2, RoundingMode.HALF_UP).toDouble())
    }

    @Test
    fun `updateBalance should calculate interest for premium plan`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(3, "premium", 1000.00, 46)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(BigDecimal(1004.17).setScale(2, RoundingMode.HALF_UP).toDouble())
    }

    @Test
    fun `updateBalance should not calculate interest for less than 30 days`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(4, "basic", 1000.00, 29)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(1000.00)
    }

    @Test
    fun `updateBalance should not calculate interest for student plan less than 366 days`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(5, "student", 1000.00, 365)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(BigDecimal(1002.50).setScale(2).toDouble())
    }

    @Test
    fun `updateBalance should not calculate interest for premium plan less than 45 days`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(6, "premium", 1000.00, 44)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(1000.00)
    }

    @Test
    fun `updateBalance should calculate interest for multiple deposits`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(7, "basic", 1000.00, 45),
            TimeDeposit(8, "student", 1000.00, 365),
            TimeDeposit(9, "premium", 1000.00, 46)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(BigDecimal(1000.83).setScale(2, RoundingMode.HALF_UP).toDouble())
        assertThat(plans[1].balance).isEqualTo(BigDecimal(1002.50).setScale(2, RoundingMode.HALF_UP).toDouble())
        assertThat(plans[2].balance).isEqualTo(BigDecimal(1004.17).setScale(2, RoundingMode.HALF_UP).toDouble())
    }

    @Test
    fun `updateBalance should handle zero balance`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(10, "basic", 0.00, 45)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(0.00)
    }

    @Test
    fun `updateBalance should handle negative balance`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(11, "basic", -1000.00, 45)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(BigDecimal(-1000.83).setScale(2, RoundingMode.HALF_UP).toDouble())
    }

    @Test
    fun `updateBalance should handle zero days`() {
        val calc = TimeDepositCalculator()
        val plans = listOf(
            TimeDeposit(12, "basic", 1000.00, 0)
        )
        calc.updateBalance(plans)
        assertThat(plans[0].balance).isEqualTo(1000.00)
    }
}